src/script/ismine.cpp

```diff
@@ -53,12 +65,35 @@ isminetype IsMine(const CKeyStore &keystore, const CScript& scriptPubKey)
         break;
     case TX_PUBKEY:
         keyID = CPubKey(vSolutions[0]).GetID();
+        if (sigversion != SIGVERSION_BASE && vSolutions[0].size() != 33) {
+            isInvalid = true;
+            return ISMINE_NO;
+        }
         if (keystore.HaveKey(keyID))
             return ISMINE_SPENDABLE;
         break;
-    case TX_PUBKEYHASH:
     case TX_WITNESS_V0_KEYHASH:
+    {
+        if (!keystore.HaveCScript(CScriptID(CScript() << OP_0 << vSolutions[0]))) {
+            // We do not support bare witness outputs unless the P2SH version of it would be
+            // acceptable as well. This protects against matching before segwit activates or
+            // uncompressed pubkey matching. This also applies to the P2WSH case.
```

Comment:
If the P2SH version of a witness program is unknown, we don't consider that is "mine". That means the user has to explicitly import the P2SH version with importaddress or addwitnessaddress, and the latter is not disabled by default before segwit activation. So it is not possible to confuse users by picking random pubkey on the blockchain and send some dust to a corresponding witness output, which is unspendable before segwit activation.

But I think now this is not for uncompressed key protection. The uncompressed key comment is just a leftover of an earlier version